### PR TITLE
Do not buffer entire body in Http Cache

### PR DESCRIPTION
**Pull request by BoD**
See #3981

### CODE REVIEW

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt [hunk 1]
```diff
@@ -1,20 +1,31 @@
...
+  fun write(response: HttpResponse, cacheKey: String)
+
   /**
    * Store the [response] with the given [cacheKey] into the cache.
-   * Note: the response's body is not consumed nor closed.
+   * A new [HttpResponse] is returned whose body, when read, will write the contents to the cache.
+   * The response's body is not consumed nor closed.
    */
-  fun write(response: HttpResponse, cacheKey: String)
+  fun writeIncremental(response: HttpResponse, cacheKey: String): HttpResponse {
```
Comment (BoD) [inline apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt:20]: A bit unfortunate. Couldn't think of another way to do this.

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt [hunk 2]
```diff
@@ -104,6 +137,51 @@ class DiskLruHttpCache(private val fileSystem: FileSystem, private val directory
...
+      val buffer = Buffer()
+      val read = try {
+        originalBody.read(buffer, byteCount)
+      } catch (e: Exception) {
+        cacheEditor.abort()
+        throw e
+      }
+
+      if (read == -1L) {
+        // We've read fully, commit the cache edit
```
Comment (BoD) [inline apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:161]: Not 100% sure this is needed.

File: tests/defer/src/jvmTest/kotlin/test/DeferJvmTest.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        """{"data":{"isColor":true},"path":["computers",1,"screen"],"hasNext":false,"label":"a"}""",
+    )
+
+    val delay = 200L
+    mockServer.enqueueMultipart(jsonList, chunksDelayMillis = delay)
+
+    var lastEmitTime = currentTimeMillis()
+    apolloClient.query(WithFragmentSpreadsQuery()).toFlow().collect {
+      // Allow a 10% margin for inaccuracies
+      assertTrue(currentTimeMillis() - lastEmitTime >= delay / 1.1)
```
Comment (martinbonnin) [inline tests/defer/src/jvmTest/kotlin/test/DeferJvmTest.kt:50]: It'd be nice to have a lower level `MockServer` API where we enqueue the chunks one after the other and read them as we go. That'd remove the potentiality of flaky timing issues.

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt (hunk 1 above)
Comment (martinbonnin) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt:20]: Is that even backward compatible? If someone implements `ApolloHttpCache` using `3.2.2` and ends up using `3.3.0` at runtime, what's going to happen? 

The Kotlin bytecode fir [CachingHttpInterceptor](https://github.com/apollographql/apollo-kotlin/blob/4b75f69669c26c0b3367eac0f50fab9dbd816a63/apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/CachingHttpInterceptor.kt#L113) shows 
```
    INVOKEVIRTUAL com/apollographql/apollo3/cache/http/DiskLruHttpCache.writeIncremental (Lcom/apollographql/apollo3/api/http/HttpResponse;Ljava/lang/String;)Lcom/apollographql/apollo3/api/http/HttpResponse;
```

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt [hunk 4]
```diff
@@ -104,6 +137,51 @@ class DiskLruHttpCache(private val fileSystem: FileSystem, private val directory
     }
   }
 
+  /**
+   * A [BufferedSource] that writes to the given cache sink as it is read.
+   */
+  private class WriteToCacheSource(
+      private val originalBody: Source,
+      private val cacheSink: Sink,
```
Comment (martinbonnin) [inline apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:145]: Nit: what about proxy? 

```suggestion
  private class ProxySource(
      private val originalSource: Source,
      private val sink: Sink,
```

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt [hunk 5]
```diff
@@ -104,6 +137,51 @@ class DiskLruHttpCache(private val fileSystem: FileSystem, private val directory
...
+      private val cacheEditor: DiskLruCache.Editor,
+  ) : Source by originalBody {
```
Comment (martinbonnin) [inline apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:147]: [Interface delegation is one of Andrey Breslav regrets in Kotlin design](https://discuss.kotlinlang.org/t/what-is-wrong-about-implementation-by-delegation/19873) so I'd prefer avoiding it. 

Especially in that case, looks like the only thing we need to proxy is `timeout`?

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt (hunk 4 above)
Comment (martinbonnin) [inline apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:145]: Nit: what about proxy? 

```suggestion
   * A [Source] that writes to the given cache sink as it is read.
   */
   private class ProxySource(
      private val originalSource: Source,
      private val sink: Sink,
```

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt (hunk 2 above)
Comment (martinbonnin) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:161]: The `DeferJvmTest` tests still pass for me without the `closeAndCommitCache()` can you add one that fails?

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt [hunk 6]
```diff
@@ -104,6 +137,51 @@ class DiskLruHttpCache(private val fileSystem: FileSystem, private val directory
...
+
+    private var hasClosedAndCommitted: Boolean = false
+
+    override fun read(sink: Buffer, byteCount: Long): Long {
+      val buffer = Buffer()
```
Comment (martinbonnin) [inline apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:152]: Sounds like this could be optimized at some point but we can always see that later ™️ :)

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt (hunk 1 above)
Comment (BoD) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt:20]: After looking a bit more at what we do with this interface:
- `CachingHttpInterceptor` instantiates `DiskLruHttpCache` directly, we don't provide a way to provide another implementation
- the `ApolloClient.httpCache: ApolloHttpCache` extension exposes the interface, but I think the only useful use-cases is to call `clearAll()` on it?

Could that mean that the interface is not really needed, and we could instead expose an extension `fun ApolloClient.clearHttpCache()`?

If yes we could:
- mark the whole interface as `@Deprecated`
- not implement it in `DiskLruHttpCache`
- mark `ApolloClient.httpCache` as deprecated, and make it return an implementation that only has `clearAll` working and the other ones being no-op

File: tests/defer/src/jvmTest/kotlin/test/DeferJvmTest.kt (hunk 3 above)
Comment (BoD) [reply tests/defer/src/jvmTest/kotlin/test/DeferJvmTest.kt:50]: Oh yeah 😊  I'll have a look!

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt [hunk 7]
```diff
@@ -104,6 +137,51 @@ class DiskLruHttpCache(private val fileSystem: FileSystem, private val directory
...
+        throw e
+      }
+
+      if (read == -1L) {
+        // We've read fully, commit the cache edit
+        closeAndCommitCache()
+        return -1L
+      }
+      buffer.peek().readAll(cacheSink)
+      sink.writeAll(buffer)
```
Comment (martinbonnin) [inline apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:166]: Sounds like something that can fail too? Should we abort the editor in that case?

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt (hunk 4 above)
Comment (BoD) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:145]: I hesitated naming it more "generically" but because of the `cacheEditor` it's pretty specific to the cache - OTOH  `ProxySource` is pretty clear too 👍

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt (hunk 1 above)
Comment (martinbonnin) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt:20]: > Could that mean that the interface is not really needed

I guess users could still call `write()` manually to warm the cache programmatically

> return an implementation that only has clearAll working and the other ones being no-op

That will make the above (hopefully very rare) use case fail so at this point, we might as well add the `HttpResponse` return value. The only thing that could break is if someone calls `ApolloHttpCache.write()` manually (very unlikely) and relies on the return value being `Unit` (even more unlikely)

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt (hunk 4 above)
Comment (martinbonnin) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:145]: Your call. The only thing I'm pretty positive about is that it is a `Source` and not a `BufferedSource`

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt (hunk 1 above)
Comment (BoD) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ApolloHttpCache.kt:20]: A lot simpler 😅  That makes sense 👍

File: apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerCommon.kt [hunk 8]
```diff
@@ -68,18 +71,64 @@ suspend fun writeResponse(sink: BufferedSink, mockResponse: MockResponse, versio
 class MockResponse(
     val statusCode: Int = 200,
     val body: ByteString = ByteString.EMPTY,
-    val chunks: List<MockChunk> = emptyList(),
+    chunks: List<MockChunk> = emptyList(),
     val headers: Map<String, String> = emptyMap(),
     val delayMillis: Long = 0,
+    private val waitForMoreChunks: Boolean = false,
 ) {
+  var chunks: List<MockChunk> = chunks
```
Comment (BoD) [inline apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerCommon.kt:79]: In addition to the channel, still keep the List, as it is useful to do asserts in tests

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt (hunk 2 above)
Comment (martinbonnin) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:128]: Did we ever find the reason behind this?

Comment (BoD) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:128]: Sorry I thought your previous comment was about the test passing without committing because it wasn't checking the presence of data in the cache - but now I'm wondering if you meant something else 😅

Comment (martinbonnin) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:128]: Let me try again

Comment (martinbonnin) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:128]: Still the case. What I'm doing is I'm commenting out the `closeAndCommitCache()` on line 129. And then run tests:

```
./gradlew -p tests build
```

And everything is still ✅  leaving me to figure out what this is for? Or do I need to run other tests?

Comment (BoD) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:128]: Ah got it (thought you were commenting the inside of `closeAndCommitCache`).

This path can be reached when an incomplete response is fully read: the JSON reader pulls more bytes, but there aren't any, so `originalSource.read` returns -1. In that case this bit of code commits early, but:

1. it's actually not useful, because `close` will still be called after a parse exception occurs (which commits at that point)
2. not 100% sure if we actually want to commit in that case? On the one hand, this would be consistent with the current behavior (currently, the data is cached before it's parsed). On the other hand, if we know for sure that -1 here means an incomplete response, it may be nice to avoid caching that?

WDYT?

Comment (martinbonnin) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:128]: Fell into this rabbit hole and turns out it's not too hard to implement. I gave it a try there: https://github.com/apollographql/apollo-kotlin/pull/4087. This is wildly untested. Feel free to include in this PR or I'll rebase and open a follow up PR later

Comment (BoD) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:128]: Yes this looks good! 🙏  Merging into this one and I'll add 2 tests for it.

Comment (BoD) [reply apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/DiskLruHttpCache.kt:128]: hmm 2 tests failed - I'll have a look before merging :P

File: apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerExtensions.kt [hunk 9]
```diff
@@ -47,25 +47,46 @@ fun createMultipartMixedChunkedResponse(
     responseDelayMillis: Long = 0,
     chunksDelayMillis: Long = 0,
     boundary: String = "-",
+    waitForMoreChunks: Boolean = false,
```
Comment (martinbonnin) [inline apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerExtensions.kt:50]: That's a lot of constructor arguments. I think MockResponse is due for a little refactoring. 

We could make it take a `Source` as body:

```kotlin
class MockResponse(
    val statusCode: Int = 200,
    val body: Source = Buffer(),
    val headers: Map<String, String> = emptyMap(),
)
```

Then `delay` and `chunks` "just" become specific `Source` implementations:

```kotlin
class DelayedSource(
    val delayMillis: Long,
    val wrappedSource:Source
    ): Source{
  private var firstRead = true

  override fun close() {
    wrappedSource.close()
  }

  override fun read(sink: Buffer, byteCount: Long): Long {
    if (firstRead) {
      // TODO: see if we can implement this more efficiently with timeouts
      usleep(delayMillis*1000)
    }
    return wrappedSource.read(sink, byteCount)
  }

  override fun timeout(): Timeout {
    TODO("Not yet implemented")
  }
}

class ChunkedSource(): Source {
  // ...
}
```

File: apollo-mpp-utils/src/commonMain/kotlin/com/apollographql/apollo3/mpp/utils.kt [hunk 10]
```diff
@@ -1,7 +1,13 @@
+@file:JvmName("-utils")
+
 package com.apollographql.apollo3.mpp
 
+import kotlin.jvm.JvmName
+
 expect fun currentTimeMillis(): Long
+expect fun currentTimeFormatted(): String
```
Comment (martinbonnin) [inline apollo-mpp-utils/src/commonMain/kotlin/com/apollographql/apollo3/mpp/utils.kt:8]:
```suggestion
/**
 * The current time as a human readable String. Used for debugging
 */
@ApolloInternal
expect fun currentTimeFormatted(): String
```

File: apollo-mpp-utils/src/appleMain/kotlin/com/apollographql/apollo3/mpp/utils.kt [hunk 11]
```diff
@@ -11,10 +13,24 @@ actual fun currentTimeMillis(): Long {
...
+
 actual fun currentThreadId(): String {
   return pthread_self()?.rawValue.toString()
 }
 
+actual fun currentThreadName(): String {
+  return if (NSThread.isMainThread) {
+    "main"
+  } else {
+    currentThreadId()
```
Comment (martinbonnin) [inline apollo-mpp-utils/src/appleMain/kotlin/com/apollographql/apollo3/mpp/utils.kt:30]: Could you use [pthread_get_name_np](https://linux.die.net/man/3/pthread_getname_np) instead? Does iOS put anything relevant there?

Comment (BoD) [reply apollo-mpp-utils/src/appleMain/kotlin/com/apollographql/apollo3/mpp/utils.kt:30]: Sadly, this returns an empty string (same for `NSThread.currentThread.name`)!

File: apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerExtensions.kt (hunk 9 above)
Comment (martinbonnin) [reply apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerExtensions.kt:50]: Could we do some kind of non-blocking/polling I/O with okio? have `read()` return zero in the source implementation while waiting and something >0 when data is available again?

Comment (BoD) [reply apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerExtensions.kt:50]: So I think 0 would make callers call again immediately, in a loop, but we could sleep for a few ms to not hog the cpu? Not ideal but maybe enough for tests?

Comment (martinbonnin) [reply apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerExtensions.kt:50]: Right. `okio.Source` is not the good abstraction then. But we could write our own `Source` abstraction with coroutines support to give a chance to JS land to "sleep"/"delay"

Comment (BoD) [reply apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerExtensions.kt:50]: What about `Flow<ByteString>`?

```
class MockResponse(
    val statusCode: Int = 200,
    val body: Flow<ByteString> = emptyFlow(),
    val headers: Map<String, String> = emptyMap(),
)
```

Example usage:

```
fun MockServer.enqueue(string: String, delayMs: Long = 0) {
  enqueue(MockResponse(
      statusCode = statusCode,
      body = flow {
        delay(delayMillis)
        emit(body.encodeUtf8())
      },
      headers = mapOf("Content-Length" to string.length.toString()),
  ))
}
```

```
val chunks = Channel<String>(Channel.UNLIMITED)
delay(200)
chunks.send("chunk 1")
delay(200)
chunks.send("chunk 2")
chunks.close()
val body: Flow<ByteString> = chunks.receiveAsFlow().map { it.encodeUtf8() }
```

Comment (BoD) [reply apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerExtensions.kt:50]: If you're OK with it I can remove the MockServer changes from this PR and open a dedicated one 😅

Comment (martinbonnin) [reply apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerExtensions.kt:50]: Sure thing!
